ckan/logic/validators.py

```diff
@@ -282,12 +319,11 @@ def resource_id_exists(value: Any, context: Context) -> Any:
     return value
 
 
-def resource_id_validator(value: Any) -> Any:
-    pattern = re.compile("[^0-9a-zA-Z _-]")
-    if pattern.search(value):
-        raise Invalid(_('Invalid characters in resource id'))
-    if len(value) < 7 or len(value) > 100:
-        raise Invalid(_('Invalid length for resource id'))
+def uuid_validator(value: Any) -> Any:
+    try:
+        uuid.UUID(value, version=4)
```

Comment:
UUIDs are case-insensitive. How about returning `str(uuid.UUID(value, version=4))` to normalize the value? In this way, we'll avoid creating entries with the same UUID written in different cases (we have a similar problem for user emails)

Reply:
:100: this. A decade ago we imported an organization list with uppercase UUIDs and now they're mixed with new orgs created with lowercase UUIDs. So untidy.

Reply:
This makes sense, but if we change the value returned we might change the case of some existing upper case UUIDs when editing objects, wouldn't that cause potential issues?

Reply:
To add to this, this would only be an issue for resources and extras (both use the same schema for creates and updates for reasons explained above). The rest of entities don't use `uuid_validator` on updates. So if an old system created resource ids like `13857D04-A35B-4834-93F1-5DD98D401D44`, when updating the resource in the new CKAN version it would change to `13857d04-a35b-4834-93f1-5dd98d401d44`. Would that be an issue assuming it is mentioned in the changelog?

Reply:
Confirmed, Bad Things™ happen when running this code in existing upper case resource ids:

```
ckanapi action package_create name=test_resource owner_org=test

ckanapi action resource_create package_id=test_resource id=C246F35C-7EAE-4F05-B971-99ED3A878B07

# Switch to code running with the uuid_validator version normalizing the value:

ckanapi action resource_update id=C246F35C-7EAE-4F05-B971-99ED3A878B07 description=test

ERROR [ckan.logic] Resource C246F35C-7EAE-4F05-B971-99ED3A878B07 exists but it is not found in the package it should belong to.
[...]
  File "/home/adria/dev/pyenvs/ckan-py3.9/ckan/ckan/logic/action/update.py", line 119, in resource_update                                 
    resource = _get_action('resource_show')(context, {'id': id})     
  File "/home/adria/dev/pyenvs/ckan-py3.9/ckan/ckan/logic/__init__.py", line 578, in wrapped                                              
    result = _action(context, data_dict, **kw)                       
  File "/home/adria/dev/pyenvs/ckan-py3.9/ckan/ckan/logic/action/get.py", line 1098, in resource_show                                     
    raise NotFound(_('Resource was not found.'))                     
ckan.logic.NotFound: Resource was not found.  
```

Looking at the DB, the existing upper case id resource was deleted and a new resource was created for the lower case version. There's a `resource_show` call at the end of `resource_update` which fails because it tries to search for the old upper case id (which is now deleted) in the dataset dict and fails.

So all in all, I don't think we can introduce this change with an extra database migration that lower cases all ids, and even then we might break external systems that rely on the existing upper case ids (although this might be fine if we document it)

Reply:
So we're not going ahead with normalizing UUIDs for this iteration, right?

Reply:
I think that if we want to normalize UUIDs we need to introduce a migration that changes all existing UUIDS to lowercase, or at least the resource ones to prevent the error I described, and clearly document it. I don't think its worth doing it given the potential for breakages. But what do you think?

Reply:
Agreed, that clean up can be part of a future release. Maybe migrate our columns to the proper UUID type at the same time for better performance.